### PR TITLE
Add VALIDATE for checking scene item invariants (code-as-documentation)

### PR DESCRIPTION
**Pull request by lukego**
I've started reading the kons-9 source to understand the data structures. I wanted a way to write notes that are shareable, useful, and maintainable. This branch is what I came up with.

Youtube one-minute summary of the motivation: https://youtu.be/yDfMQ3Q9gFA

Or in code,

- 84662e4 adds one-liner docstrings and types to each slot of POINT-CLOUD, CURVE, and POLYHEDRON. Maybe this information is obvious to some but it wasn't to me!
- 8fd2c00 specifies invariant properties of these same classes in the form of VALIDATE methods. You can use them to check your scenes for corrupt data.
- 3de0f87 exercises VALIDATE in the testsuite with both valid and invalid example scene items.

Whaddayareckon? Is working through the data structures and documenting them this way a useful form of code-as-documentation?

CI should confirm that these tests pass:

```
CL-USER> (kons-9/testsuite::validate/valid)
Name: KONS-9/TESTSUITE:VALIDATE/VALID
Total: 15
Success: 15/15 (100%)
Failure: 0/15 (0%)
Condition: 0/15 (0%)
Outcome: Success
...
CL-USER> (kons-9/testsuite::validate/invalid)
Name: KONS-9/TESTSUITE:VALIDATE/INVALID
Total: 9
Success: 9/9 (100%)
Failure: 0/9 (0%)
Condition: 0/9 (0%)
Outcome: Success
...
```

### CODE REVIEW

**Comment by lukego** [comment]
I don't know why CI failed this branch. The same Github Action gave it a green check-mark pass on my lukego/kons-9 fork: https://github.com/lukego/kons-9/commits/validate. Spurious?

File: src/kernel/validate.lisp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  (let ((errors 0))
+    (handler-bind ((validation-error (lambda (c)
+                                       (declare (ignore c))
+                                       (incf errors)
+                                       (invoke-restart 'ignore))))
+      (validate scene-item)
+      errors)))
+
+(defun verror (format &rest format-args)
+  "Signal a (continuable) validation error on *VALIDATION-SUBJET*."
```
Comment (kaveh808) [inline src/kernel/validate.lisp:59]: typo: SUBJECT

File: src/kernel/color.lisp [hunk 2]
```diff
@@ -2,6 +2,14 @@
 
 ;;;; color ==============================================================
 
+(deftype color ()
+  "Mutable vector of R,G,B,A signle-float values between zero and one."
```
Comment (kaveh808) [inline src/kernel/color.lisp:6]: typo: single-float

**Comment by foretspaisibles** [comment]
> I don't know why CI failed this branch. The same Github Action gave it a green check-mark pass on my lukego/kons-9 fork: https://github.com/lukego/kons-9/commits/validate. Spurious?

I add a look at the logs and MacPorts failed to install sbcl (“port not found” which is odd…). This should be a temporary failure but this does not make a lot sense to me. Let's see if restarting works.

**Comment by foretspaisibles** [comment]
That's very nice to see more tests and presenting the idea with a short video is also very pleasant! For the future we could investigate the idea of writing a specific `define-assertion` rather than counting the number of error reports, so that we have the chance to produce a more detailed failure explanation when the test fail.

The file `testsuite/assertions.lisp` illustrates how to develop custom assertions and have display a detailed report with custom information when they fail.

**Comment by foretspaisibles** [comment]
> I don't know why CI failed this branch. The same Github Action gave it a green check-mark pass on my lukego/kons-9 fork: https://github.com/lukego/kons-9/commits/validate. Spurious?

Definitely spurious, I restarted the workflow and tests are now green. It looks like GitHub Actions tripped on their own foot there.

**Comment by foretspaisibles** [comment]
I just took a couple of minutes to fix the typo you have spot a few days ago @kaveh808!

**Comment by lukego** [comment]
Thanks. I'll reopen this PR when it is more complete. If anyone wants to take it and run with it then that's fine with me though.

**Comment by kaveh808** [comment]
Thanks @lukego. I'm taking a bit of a hands-off approach to the validation, and happy to let you and (I think?) @foretspaisibles run with it.

**Pull request closed without merging**